src/java.base/share/classes/sun/nio/ch/NativeSocketAddress.java

```diff
@@ -47,63 +61,97 @@
 
-    private static final int SIZEOF_SOCKADDR4     = sizeofSockAddr4();
-    private static final int SIZEOF_SOCKADDR6     = sizeofSockAddr6();
-    private static final int SIZEOF_SOCKETADDRESS = Math.max(SIZEOF_SOCKADDR4, SIZEOF_SOCKADDR6);
-    private static final int SIZEOF_FAMILY        = sizeofFamily();
-    private static final int OFFSET_FAMILY        = offsetFamily();
-    private static final int OFFSET_SIN4_PORT     = offsetSin4Port();
-    private static final int OFFSET_SIN4_ADDR     = offsetSin4Addr();
-    private static final int OFFSET_SIN6_PORT     = offsetSin6Port();
-    private static final int OFFSET_SIN6_ADDR     = offsetSin6Addr();
-    private static final int OFFSET_SIN6_SCOPE_ID = offsetSin6ScopeId();
-    private static final int OFFSET_SIN6_FLOWINFO = offsetSin6FlowInfo();
+        // IPv4 address field's layout path
+        PathElement[] saddr_layout_path = switch (OperatingSystem.current()) {
+            case WINDOWS -> new PathElement[]{
+                    groupElement("sin_addr"),
+                    groupElement("S_un"),
+                    groupElement("S_addr")
+            };
+            case LINUX, MACOS -> new PathElement[]{
+                    groupElement("sin_addr"),
+                    groupElement("s_addr")};
+            case AIX -> throw new RuntimeException("AIX not supported yet");
+        };
+        var saddrVH = sockaddr_in.layout().varHandle(saddr_layout_path);
+        // pos=0 is MemorySegment; pos=1 is the offset
+        INET4_S_ADDR_VH = MethodHandles.insertCoordinates(saddrVH, 1, 0L);
+    }
 
-    // SOCKETADDRESS
-    private final long address;
+    private final MemorySegment memory;
 
     long address() {
-        return address;
+        return memory.address();
     }
 
-    NativeSocketAddress() {
-        long base = UNSAFE.allocateMemory(SIZEOF_SOCKETADDRESS);
-        UNSAFE.setMemory(base, SIZEOF_SOCKETADDRESS, (byte) 0);
-        this.address = base;
+    private NativeSocketAddress(MemorySegment memory) {
+        if (memory.byteSize() != SIZEOF_SOCKETADDRESS) {
+            throw new IllegalArgumentException();
+        }
+        this.memory = memory;
     }
 
     /**
      * Allocate an array of native socket addresses.
      */
-    static NativeSocketAddress[] allocate(int count) {
+    static NativeSocketAddress[] allocate(int count, Arena arena) {
+        MemorySegment addressesMemory = arena.allocate(SIZEOF_SOCKETADDRESS * count);
+
         NativeSocketAddress[] array = new NativeSocketAddress[count];
         for (int i = 0; i < count; i++) {
             try {
-                array[i] = new NativeSocketAddress();
+                MemorySegment addressMemory = addressesMemory.asSlice(
+                        i * SIZEOF_SOCKETADDRESS, SIZEOF_SOCKETADDRESS);
+                array[i] = new NativeSocketAddress(addressMemory);
             } catch (OutOfMemoryError e) {
-                freeAll(array);
+                freeAll(arena);
```

Comment:
An upside of doing just one allocation and then split it up in three parts is that we do not have to handle the OOME. Either the one and only allocation succeeds or it fails. No need for explicitly freeing memory.